server/mdm/lifecycle/lifecycle.go

```diff
@@ -0,0 +1,272 @@
+package mdmlifecycle
+
+import (
+	"context"
+
+	"github.com/fleetdm/fleet/v4/server/contexts/ctxerr"
+	"github.com/fleetdm/fleet/v4/server/contexts/license"
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	"github.com/fleetdm/fleet/v4/server/worker"
+	kitlog "github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+)
+
+// HostAction is a supported MDM lifecycle action that can be performed on a
+// host.
+type HostAction string
+
+const (
+	// HostActionTurnOn performs tasks right after a host turns on MDM.
+	HostActionTurnOn HostAction = "turn-on"
+	// HostActionTurnOn performs tasks right after a host turns off MDM.
+	HostActionTurnOff HostAction = "turn-off"
+	// HostActionTurnOn perform tasks to reset mdm-related information.
+	HostActionReset HostAction = "reset"
+	// HostActionDelete perform tasks to cleanup MDM information when a
+	// host is deleted from fleet.
+	HostActionDelete HostAction = "delete"
+)
+
+// HostOptions are the options that can be provided for an action.
+//
+// Not all options are required for all actions, each individual action should
+// validate that it receives the required information.
+type HostOptions struct {
+	Action          HostAction
+	Platform        string
+	UUID            string
+	HardwareSerial  string
+	HardwareModel   string
+	EnrollReference string
+	Host            *fleet.Host
+}
+
+// HostLifecycle manages MDM host lifecycle actions
+type HostLifecycle struct {
+	ds     fleet.Datastore
+	logger kitlog.Logger
+}
+
+// New creates a new HostLifecycle struct
+func New(ds fleet.Datastore, logger kitlog.Logger) *HostLifecycle {
+	return &HostLifecycle{
+		ds:     ds,
+		logger: logger,
+	}
+}
+
+// Do executes the provided MDMHostLifecycleAction based on the platform requested
```

Comment:
Looks like this refers to the type of opts.Action? If so:

```suggestion
// Do executes the provided HostAction based on the platform requested
```